### PR TITLE
`sage.rings.polynomial.laurent_polynomial_ring_base`: Split out from `.laurent_polynomial_ring`

### PR DESCRIPTION
**Pull request by mkoeppe**
### 📚 Description
We move the abstract base class `LaurentPolynomialRing_generic` to a separate module.
This is for modularization purposes (meta-ticket #32414).

We also deprecate the function `is_LaurentPolynomialRing` and replace the (few) uses by `isinstance`.

### 📝 Checklist

- [x] I have made sure that the title is self-explanatory and the description concisely explains the PR.
- [x] I have linked an issue or discussion.
- [ ] I have created tests covering the changes.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: src/sage/rings/polynomial/laurent_polynomial_ring_base.py [hunk 1]
```diff
@@ -0,0 +1,535 @@
...
+r"""
+Ring of Laurent Polynomials (base class)
+
+If `R` is a commutative ring, then the ring of Laurent polynomials in `n`
+variables over `R` is `R[x_1^{\pm 1}, x_2^{\pm 1}, \ldots, x_n^{\pm 1}]`.
+We implement it as a quotient ring
+
+.. MATH::
+
+    R[x_1, y_1, x_2, y_2, \ldots, x_n, y_n] / (x_1 y_1 - 1, x_2 y_2 - 1, \ldots, x_n y_n - 1).
```
Comment (tscrim) [inline src/sage/rings/polynomial/laurent_polynomial_ring_base.py]: This is not true: It is not in the category of quotients, which reflects that it doesn't have methods like `ambient()` and `lift()`, nor does it reflect the internal structure. We have implemented it as a localization. Although it is fine to say it is isomorphic to this quotient ring; not that anything needs to be said here.

**Review by tscrim — CHANGES_REQUESTED** [review]
As I understand it, the rest is just moving stuff line-for-line, which is okay. I know this is for modularization purposes, but it looks weird that the main global entry point is separate from the generic class. I would consider changing the generic class name to `LaurentPolynomialRing` and moving the dispatch function into its `__classcall_private__`.

File: src/sage/rings/polynomial/laurent_polynomial_ring_base.py (hunk 1 above)
Comment (mkoeppe) [reply src/sage/rings/polynomial/laurent_polynomial_ring_base.py]: Not my text

**Review by tscrim — APPROVED** [review]
Thank you. LGTM.

**Comment by mkoeppe** [comment]
Thanks!

**Review by tscrim — APPROVED** [review]

**Comment by mkoeppe** [comment]
Thanks again!

**Comment by vbraun** [comment]
```
**********************************************************************
File "src/sage/interfaces/r.py", line 697, in sage.interfaces.r.R.__reduce__
Failed example:
    rlr, t = r.__reduce__()  # optional - rpy2
Exception raised:
    Traceback (most recent call last):
      File "/home/release/Sage/src/sage/doctest/forker.py", line 695, in _run
        self.compile_and_execute(example, compiler, test.globs)
      File "/home/release/Sage/src/sage/doctest/forker.py", line 1093, in compile_and_execute
        exec(compiled, globs)
      File "<doctest sage.interfaces.r.R.__reduce__[0]>", line 1, in <module>
        rlr, t = r.__reduce__()  # optional - rpy2
                 ^^^^^^^^^^^^^^
      File "/home/release/Sage/local/var/lib/sage/venv-python3.11.1/lib/python3.11/copyreg.py", line 76, in _reduce_ex
        raise TypeError(f"cannot pickle {cls.__name__!r} object")
    TypeError: cannot pickle 'LazyImport' object
**********************************************************************
```

**Comment by mkoeppe** [comment]
This one is also unrelated.

**Pull request merged**